### PR TITLE
referral tracking

### PR DESCRIPTION
**Pull request by megaherz**

### CODE REVIEW

File: internal/service/service.go [hunk 1]
```diff
@@ -90,7 +90,16 @@ func (s *service) Register(ctx context.Context, email, address string) error {
 		return fmt.Errorf("failed to send email: %w", err)
 	}
 
-	if err := s.storage.UpsertRequest(ctx, owner, email, address, code); err != nil {
+	var referralCodeAsNullString sql.NullString
+	if referralCode != nil {
+		// check the given referral code exists
+		if _, err := s.storage.GetRequestByOwnReferralCode(ctx, *referralCode); err != nil {
+			return fmt.Errorf("failed to get request by own referral code: %w", err)
```
Comment (mikluke) [inline internal/service/service.go:97]: You've missed check for storage.ErrReferralCodeNotFound

File: internal/storage/postgres/postgres.go [hunk 2]
```diff
@@ -78,22 +96,38 @@ func (p pg) UpsertRequest(ctx context.Context, owner, email, address, code strin
...
-func (p pg) CreateReferral(ctx context.Context, referral *storage.Referral) error {
-	if _, err := p.db.NamedExecContext(ctx,
-		`INSERT INTO referral (sender, receiver, registered_at) VALUES (:sender, :receiver, CURRENT_TIMESTAMP)`,
-		referral); err != nil {
-		if isUniqueViolationErr(err) {
-			return storage.ErrReferralExists
+func (p pg) CreateReferralTracking(ctx context.Context, receiver string, referralCode string) error {
+	if _, err := p.db.ExecContext(ctx,
+		`INSERT INTO referral_tracking (sender, receiver, registered_at) 
+                   VALUES (
```
Comment (mikluke) [inline internal/storage/postgres/postgres.go:102]: formatting

**Review by mikluke — APPROVED** [review]

File: internal/storage/postgres/postgres.go (hunk 2 above)
Comment (megaherz) [reply internal/storage/postgres/postgres.go:102]: Fixed

File: scripts/migrations/postgres/20211109185301_referral.up.sql [hunk 3]
```diff
@@ -1,7 +1,8 @@
 CREATE EXTENSION IF NOT EXISTS "pgcrypto";
 
-ALTER TABLE request ADD COLUMN referral_code TEXT NULL;
-CREATE UNIQUE INDEX request_referral_code_idx ON request(referral_code);
+ALTER TABLE request
+    ADD COLUMN own_referral_code TEXT NULL UNIQUE,
+    ADD COLUMN registered_by_referral_code TEXT NULL REFERENCES request (own_referral_code);
```
[the remark this answers is not in the document]
Comment (megaherz) [reply scripts/migrations/postgres/20211109185301_referral.up.sql:5]: Thx, renamed

**Pull request merged**